### DISCUSSION TITLE
Adding Checker Framework

### DISCUSSION THREAD

**Discussion by MarcinOrlowski** [logisim-evolution/logisim-evolution#1104] [Dev talk]
I am strongly considering adding [Checker Framework](https://checkerframework.org/) to our code base. But as this is not small task, let's discuss use of such tools in the project first.

**Comment by maehne** [comment]
I'm all for trying to improve type safety and using static code analysis for finding bugs in in our code base and thus improve its quality. However, I am not sure that it is worth the effort if we need to spend a lot of time to go through all of our code base to add annotations, which guide the checkers. I think we can come already a long way by addressing the warnings, which get generated by the compiler and static analysers integrated in our current IDEs and as part of the CI builds. We should also try to remove code smells when we stumble upon them, which may hide bugs, e.g., by catching any exception and doing nothing.

Making Logisim-evolution a stable/reliable program is good, but there are no lifes/sensitive data at risk if it crashes. Therefore, we don't need to strive for the same quality standards that are necessary for safety/security critical code. We also need to be sure that our source code stays readable for average Java developers, so that we keep the entry bar low for new contributors. Unfamiliar code annotations and requiring installation of additional tools besides the JDK, aren't helping in this respect. So, for the moment, I don't think adding the Checker Framework to our code base would bring us enough benefits.

**Reply by MarcinOrlowski** [reply]
I did some deeper investigation on the subject and unfortunately there some small issues with Checker disallowing gradual implementation (which is a must-have in the project of size of ours) and one big issue of Checker not understanding switch expression. So for now it is not going to happen anyway.
